lib/calculate_reference_proteome_similarity.py

```diff
@@ -0,0 +1,144 @@
+        self.output_file = output_file
+        self.metric_file = "{}.reference_matches".format(output_file)
+        self.peptide_sequence_length = peptide_sequence_length
+        self.match_length = match_length
+        self.species = species
+        self.file_type = file_type
+        self.species_to_organism = {
+            'human': 'Homo sapiens',
+            'mouse': 'Mus musculus',
+            'chimpanzee': 'Pan troglodytes',
+            'macaque': 'Macaca',
+            'cow': 'Bos taurus',
+            'horse': 'Equus caballus',
+            'pig': 'Sus scrofa',
+        }
+
+    def reference_match_headers(self):
+        return [
+            'Reference Match',
+        ]
+
+    def get_mt_peptides(self):
+        records = list(SeqIO.parse(self.input_fasta, "fasta"))
+        if self.file_type == 'vcf':
+            records_dict = {x.id.replace('MT.', ''): str(x.seq) for x in filter(lambda x: x.id.startswith('MT.'), records)}
+        else:
+            records_dict = {x.id: str(x.seq) for x in records}
+        return records_dict
+
+    def get_wt_peptides(self):
+        if self.file_type == 'vcf':
+            records = list(SeqIO.parse(self.input_fasta, "fasta"))
+            records_dict = {x.id.replace('WT.', ''): str(x.seq) for x in filter(lambda x: x.id.startswith('WT.'), records)}
+        else:
+            return {}
+        return records_dict
+
+    def extract_n_mer(self, full_peptide, subpeptide_position, mutation_position, mt_length):
+        #For non-frameshifts this ensures that we only test match_length epitopes that overlap the mutation
+        #If we extract a larger region, we will get false-positive matches against the reference proteome
+        #from the native wildtype portion of the peptide
+        flanking_sequence_length = self.match_length - 1
+        mt_start = (subpeptide_position-1) + (mutation_position-1)
+        start = mt_start - flanking_sequence_length
+        if start < 0:
+            start = 0
+        end = mt_start + mt_length + flanking_sequence_length
+        return full_peptide[start:end]
+
+    def extract_n_mer_from_fs(self, full_peptide, wt_peptide, epitope, peptide_sequence_length, subpeptide_position):
+        #For frameshifts we want to test all downstream epitopes that would be part of the peptide_sequence_length
+        #peptide since they are all potentially novel
+        if peptide_sequence_length%2 == 0:
+            flanking_sequence_length = int(peptide_sequence_length/2)
+        else:
+            flanking_sequence_length = int((peptide_sequence_length-1)/2)
+        start = subpeptide_position - 1 - flanking_sequence_length
+        if start < 0:
+            start = 0
+        #This catches cases where the start position would cause too many leading wildtype amino acids, which would result
+        #in false-positive reference matches
+        diff_position = [i for i in range(len(wt_peptide)) if wt_peptide[i] != full_peptide[i]][0]
```

Comment:
as a non-pythonista, this syntax is a little confusing. I _think_ it's returning a list of the positions where the WT and MT peptides differ, right?

Reply:
`i for i in range(len(wt_peptide)) if wt_peptide[i] != full_peptide[i]]` returns a list of positions that are different, and the subsequent `[0]` grabs the first entry, i.e. the first position that is different.